web/modules/custom/solution/src/SolutionAffiliationFieldItemList.php

```diff
@@ -102,55 +127,67 @@ protected function getAffiliation(): array {
     $solution = $this->getEntity();
-    /** @var \Drupal\sparql_entity_storage\SparqlEntityStorageGraphHandlerInterface $graph_handler */
-    $graph_handler = \Drupal::service('sparql.graph_handler');
-    $graph_uri = $graph_handler->getBundleGraphUri('rdf_entity', 'collection', $solution->get('graph')->target_id);
+    $connection = $this->getSparqlConnection();
+    $graph_uris = $this->getAvailableGraphs();
+
     /** @var \Drupal\sparql_entity_storage\SparqlEntityStorageFieldHandlerInterface $field_handler */
     $field_handler = \Drupal::service('sparql.field_handler');
     $field_uri = $field_handler->getFieldPredicates('rdf_entity', 'field_ar_affiliates')['collection'];
 
-    // Get existing affiliation.
-    $select[] = 'SELECT ?id';
-    $select[] = "FROM <{$graph_uri}>";
-    $select[] = "WHERE { ?id <{$field_uri}> <{$solution->id()}> . }";
-    $select[] = "ORDER BY (?id)";
-    $existing_ids = [];
-    foreach ($sparql->query(implode("\n", $select)) as $item) {
-      $existing_ids[] = (string) $item->id;
-    }
+    $affected_ids = [];
+    foreach ($graph_uris as $graph_uri) {
+      // Get existing affiliation.
+      $select = [];
+      $select[] = 'SELECT ?id';
+      $select[] = "FROM <{$graph_uri}>";
+      $select[] = 'WHERE {';
+      $select[] = "?id <{$field_uri}> <{$solution->id()}> .";
+      // Ensure the entity exists in the graph.
+      $select[] = "?id a ?type .";
+      $select[] = '}';
+      $select[] = "ORDER BY (?id)";
+
+      $existing_ids = [];
+      foreach ($connection->query(implode("\n", $select)) as $item) {
+        $existing_ids[] = (string) $item->id;
+      }
 
-    $new_ids = array_map(function (EntityReferenceItem $field_item): string {
-      return $field_item->target_id;
-    }, $this->list);
-    sort($new_ids);
+      $new_ids = array_map(function (EntityReferenceItem $field_item): string {
+        return $field_item->target_id;
+      }, $this->list);
+      sort($new_ids);
 
-    // The affiliation has been preserved. Exit here.
-    if ($new_ids === $existing_ids) {
-      return;
-    }
+      // The affiliation has been preserved. Exit here.
+      if ($new_ids === $existing_ids) {
+        return;
+      }
 
-    // Delete existing affiliation.
-    $delete[] = "WITH <{$graph_uri}>";
-    $delete[] = "DELETE { ?id <{$field_uri}> <{$solution->id()}> . }";
-    $delete[] = "WHERE { ?id <{$field_uri}> <{$solution->id()}> . }";
-    $sparql->query(implode("\n", $delete));
-
-    // Insert new affiliation.
-    $insert[] = "WITH <{$graph_uri}>";
-    $insert[] = 'INSERT {';
-    foreach ($new_ids as $id) {
-      $insert[] = "  <{$id}> <{$field_uri}> <{$solution->id()}> .";
+      // Delete existing affiliation.
+      $delete = [];
+      $delete[] = "WITH <{$graph_uri}>";
+      $delete[] = "DELETE { ?id <{$field_uri}> <{$solution->id()}> . }";
+      $delete[] = "WHERE { ?id <{$field_uri}> <{$solution->id()}> . }";
+      $connection->query(implode("\n", $delete));
```

Comment:
As I performance improvement, I would run a single query after `foreach (...) {...}` across all graphs.

Reply:
I had troubles in the past with multiple `WITH` statements. It seemed that it was accepting the query as valid but was running the query across ALL graphs instead of the ones in the `WITH` restriction. If possible, I would keep it isolated to a single graph.